art/attacks/hop_skip_jump.py

```diff
@@ -0,0 +1,513 @@
+                                                     clip_min=clip_min, clip_max=clip_max, threshold=0.001)
+                    self.norm = tmp_norm
+                    initial_sample = random_img, random_class
+
+                    logging.info('Found initial adversarial image for targeted attack.')
+                    break
+            else:
+                logging.warning('Failed to draw a random image that is adversarial, attack failed.')
+
+        else:
+            # The initial image satisfied
+            if adv_init is not None and init_pred != y_p:
+                return adv_init.astype(NUMPY_DTYPE), y_p
+
+            # The initial image unsatisfied
+            for _ in range(self.init_size):
+                random_img = nprd.uniform(clip_min, clip_max, size=x.shape).astype(x.dtype)
+                random_class = np.argmax(self.classifier.predict(np.array([random_img])), axis=1)[0]
+
+                if random_class != y_p:
+                    # Binary search to reduce the l2 distance to the original image
+                    tmp_norm = self.norm
+                    self.norm = 2
+                    random_img = self._binary_search(current_sample=random_img, original_sample=x, target=y_p,
+                                                     clip_min=clip_min, clip_max=clip_max, threshold=0.001)
+                    self.norm = tmp_norm
+                    initial_sample = random_img, y_p
+
+                    logging.info('Found initial adversarial image for untargeted attack.')
+                    break
+            else:
+                logging.warning('Failed to draw a random image that is adversarial, attack failed.')
+
+        return initial_sample
+
+    def _attack(self, initial_sample, original_sample, target, clip_min, clip_max):
+        """
+        Main function for the boundary attack.
+
+        :param initial_sample: An initial adversarial example.
+        :type initial_sample: `np.ndarray`
+        :param original_sample: The original input.
+        :type original_sample: `np.ndarray`
+        :param target: The target label.
+        :type target: `int`
+        :param clip_min: Minimum value of an example.
+        :type clip_min: `float`
+        :param clip_max: Maximum value of an example.
+        :type clip_max: `float`
+        :return: an adversarial example.
+        :rtype: `np.ndarray`
+        """
+        # Set current perturbed image to the initial image
+        current_sample = initial_sample
+
+        # Main loop to wander around the boundary
+        for _ in range(self.max_iter):
+            # First compute delta
+            delta = self._compute_delta(current_sample=current_sample, original_sample=original_sample,
+                                        clip_min=clip_min, clip_max=clip_max)
+
+            # Then run binary search
+            current_sample = self._binary_search(current_sample=current_sample, original_sample=original_sample,
```

Comment:
Following lines 202 and 265: The possible new argument `norm` to `_binary_search` would be set to `self.norm` in this line.

Reply:
Solved.